modules/history/src/history-helper.js

```diff
@@ -0,0 +1,186 @@
+                const snapshotPath = snapshotState[snapshotPathKey];
+                snapshotPath.forEach(
+                    (snapshotEntity) => {
+                        Object.keys(currentState).forEach(
+                            (currentPathKey) => {
+                                const currentPath = currentState[currentPathKey];
+                                currentPath.forEach(
+                                    (currentEntity) => {
+                                        if ((currentEntity.id === snapshotEntity.id)
+                                        && (snapshotPathKey !== currentPathKey)) {
+                                            dataStore.removeEntity(currentEntity.id);
+                                            const entity = SerializableClass
+                                                .$unserialize(snapshotEntity);
+                                            dataStore.addEntity(entity, snapshotPathKey);
+                                        }
+                                    },
+                                );
+                            },
+                        );
+                    },
+                );
+            },
+        );
+    },
+
+    /**
+     * Take a snapshot and returns an object with entity.id
+     * as keys and entity as values
+     * @param {object} snapshotState
+     * @return {object} The flattened object
+     */
+    flattenSnapshot: (snapshotState) => {
+        const flattenSnapshotObject = {};
+        Object.keys(snapshotState).forEach(
+            (path) => {
+                snapshotState[path].forEach(
+                    (entity) => {
+                        flattenSnapshotObject[entity.id] = entity;
+                    },
+                );
+            },
+        );
+        return flattenSnapshotObject;
+    },
+
+    /**
+     * Check the difference between two properties of object
+     */
+    checkDiff: (property1, property2) => {
+        if (property1 === null || property2 === null
+            || typeof property1 !== "object"
+            || Object.keys(property1).length === 0
+        ) {
+            return (property1 !== property2);
+        }
+        if (Array.isArray(property1)) {
+            return historyHelper.checkDiffInArray(property1, property2);
+        }
+        return Object.keys(property1).some(
+            propertyKey => historyHelper.checkDiff(property1[propertyKey], property2[propertyKey]),
+        );
+    },
+
+    /**
+     * Check the difference between two arrays
+     * @param {array} array1
+     * @param {array} array2
+     */
+    checkDiffInArray(array1, array2) {
+        return (array1.length !== array2.length || array1.sort().some(
+            (value, index) => value !== array2.sort()[index],
+        ));
+    },
+
+    /**
+     * Quick method to clone an object to have different instance
+     * @param {object} object
+     * @return {object}
+     */
+    cloneObject(object) {
```

Comment:
There is already a `cloneDeep` in obsidian helpers. Please use it instead re-implementing it: https://github.com/wanadev/obsidianjs/blob/master/obsidian/src/helpers.js#L132

Reply:
I think we should not even use a function in that case since it's used only here,

`const cloneObject = JSON.stringify(JSON.parse(snapshotEntity)));`

Seems more straigthforward as it's a commonly used "hack"